### PR TITLE
Default `moment` for `CustomDist` provided with a `dist` function

### LINKED ISSUE #6804: Implement default `moment` function for `CustomDist` provided with a `dist` function
### Description

```python
import pymc as pm

def dist(mu, sigma, size):
  return pm.math.exp(pm.Normal.dist(mu, sigma size=size))`

with pm.Model() as m:
  ...
  x = pm.CustomDist("x", mu, sigma, dist=dist)

m.initial_point()
```

Since the `dist` is a symbolic forward graph, we could implement the `moment` automatically as
```python
def moment_dist(mu, sigma, size):
  return pm.math.exp(moment(pm.Normal.dist(mu, sigma, size))
```

Would just need to replace the RVs in the `dist` graph, by their moments.

### PR DESCRIPTION
**Pull request by aerubanov**
Add default implementation of `moment` for `CustomDist` with a `dist` function and close #6804

----
:books: Documentation preview :books:: https://pymc--6873.org.readthedocs.build/en/6873/

### CODE REVIEW

**Comment by aerubanov** [comment]
@ricardoV94 I added implementation which use graph rewriting to replace distributions by corresponding moments. Could you please take a look?

File: pymc/distributions/distribution.py [hunk 1]
```diff
@@ -83,6 +85,21 @@
...
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(ReplaceValidate())
+
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            for i, inp in enumerate(node.inputs):
+                if (
+                    hasattr(inp, "owner")
+                    and hasattr(inp.owner, "op")
+                    and isinstance(inp.owner.op, Distribution)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Maybe we should check if it's a RandomVariable instead of Distribution.

We also need to check if an Op had an InnerGraph (like Scan) and do the replacements inside

**Review by ricardoV94 — COMMENTED** [review]
Neat approach! 

Just need to think about Ops with InnerGraphs

File: pymc/distributions/distribution.py (hunk 1 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: Thank you @ricardoV94, will do it.

File: pymc/distributions/distribution.py [hunk 2]
```diff
@@ -83,6 +85,21 @@
...
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(ReplaceValidate())
+
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            for i, inp in enumerate(node.inputs):
+                if (
+                    hasattr(inp, "owner")
+                    and hasattr(inp.owner, "op")
+                    and isinstance(inp.owner.op, RandomVariable)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Ah could also be a symbolic one!

```suggestion
                    and isinstance(inp.owner.op, (RandomVariable, SymbolicRandomVariable))
```

Comment (aerubanov) [reply pymc/distributions/distribution.py]: Thank you!

**Comment by aerubanov** [comment]
@ricardoV94 Do you know som simple example of symbolic `dist` with `Scan` op? I want to add test case with `Scan`, but I do not used it before, so when I try this:
```python
def test_custom_dist_custom_moment_inner_graph(self):
        def dist(mu, size):
            ys, _ = pytensor.scan(
                fn=lambda x: pt.exp(pm.Normal.dist(x, 1, size=size)),
                sequences=[mu],
                outputs_info=[None], 
                name="ys",
                )
            return pt.sum(ys)
        
        with Model() as model:
            CustomDist("x", pt.ones(2), dist=dist)
        assert_moment_is_expected(model, 2)
```
I get  `ValueError: No update found for at least one RNG used in Scan Op Scan{ys, while_loop=False, inplace=none}.` I guess I just use `scan` wrong

**Comment by ricardoV94** [comment]
You need to return random updates from the scan. Check the utility `collect_default_updates` and the code example in the docstrings:

https://github.com/pymc-devs/pymc/blob/main/pymc/pytensorf.py#L1000

**Comment by aerubanov** [comment]
@ricardoV94 yeah, it work, thank you!

File: pymc/distributions/distribution.py [hunk 3]
```diff
@@ -83,6 +85,18 @@
...
 
 
+class MomentRewrite(GraphRewriter):
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(ReplaceValidate())
+
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            if hasattr(node.op, "fgraph"):
+                self.rewrite(node.op.fgraph)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: We have to double check this isn't mutating the fgraph of the original Ops (I am not sure they are cloned by default in FunctionGraph).

More generally I am not sure this is the best approach. The fgraph is supposed to be just a property of the Op, the real underlying variables are Op.outputs I think? So we could create a FunctionGraph from those, rewrite them and then create a new node with a new op that uses those new outputs and replace it in the outer graph...

However these Ops don't have a shared construction API IIRC, so we may need special logic for each one. They are Scan, OpFromGraph, Composite, and ScalarLoop. 

I think we can focus on Scan alone since the others aren't user facing or don't make sense inside a CustomDist (just raise NotImplementedError in those cases)

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 thank you for review! I will try to replace `Scan` op in outer graph by new node with moment replaced inner graph.

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 If I understand correctly, new `Scan` op in outer graph is already created in my implementation when I clone graph  from dist function. So we need only replace inner graph, or do I miss something?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: I mean something like
```python
if hasattr(node.op, "fgraph"):
    inner_graph = node.op.fgraph.clone()
    self.rewrite(inner_graph)
    node.op.fgraph = inner_graph
```
should be enough?

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Unfortunately we can't just modify the fgraph, as that's just a dynamic property of the Op. In some places it may be used directly, but that's not guaranteed. The Op inner inputs / outputs are the real properties of the Op, which means we need to create a new Scan op that uses the new outputs. Recreating a Scan is pretty annoying, you can check pytensor scan rewrites to see how it's approached there.

File: tests/distributions/test_distribution.py [hunk 4]
```diff
@@ -473,25 +473,22 @@ def test_custom_dist_default_moment(self, dist_params, size, expected, dist_fn):
...
-
             xs, updates = scan(
                 fn=scan_step,
-                outputs_info=[x0],
-                n_steps=2,
+                sequences=[pt.ones(2)],
                 name="xs",
             )
-            return xs[-1]
+            return pt.sum(xs)
```
Comment (ricardoV94) [inline tests/distributions/test_distribution.py]: `pt.sum` gives a variable for which PyMC doesn't know the logp. You can return the whole `xs` that should be fine. Would be good to use something like a `Uniform(-2, -1)` inside the Scan, for which the old `CustomDist` moment of zeros would fail. 

And then you can remove the `check_finite_logp=False`

File: pymc/distributions/distribution.py [hunk 5]
```diff
@@ -622,13 +679,16 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
+        def dist_moment(rv, size, *dist_params):
+            # TODO: add check for other op with inner graph (not Scan)
+            fgraph = FunctionGraph(outputs=[dist(*dist_params, size=size)], clone=True)
```
Comment (aerubanov) [inline pymc/distributions/distribution.py]: @ricardoV94 do we need some check here for` OpFromGraph`, `Composite`, and `ScalarLoop` Ops in `fgraph`? I mean to use default moment like we did before in this case.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Yes, let's raise `NotImplementedError` for those, put the new logic in a `try/except` and default back to the zeros like before when it raises. 

I think we should put this `dist_moment` helper inside the old `default_moment` function?

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Wait only `OpFromGraph` is problematic. The other remaining cases can't have `RandomVariables` inside so we can just replace their inputs outside without concern for their inner graph.

Let's only raise when it has InnerGraph and it's not Scan/Composite/ScalarLoop, which are the cases we can safely handle.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Then we can add a github issue to also implemnet the OpFromGraph case which shouldn't be too hard (similar to scan). Or do you want to do that in this PR after all? That would neatly conclude all cases.

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 Well, I think I want to implement  the OpFromGraph case in this PR. Do you know some example on how to use OpFromGraph? I need to create some test case for it

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: I'll write something tomorrow. Thanks for the initiative in advance!

**Comment by aerubanov** [comment]
@ricardoV94 just friendly reminder about example for `OpFromGraph` )

**Comment by aerubanov** [comment]
@aerubanov Yeah, I need to add test case for `pm.Truncated` because I have not it now. And may be will need to add some changes in my implementation, because I did not take into account this case when was worked on my implementation. So thank you for pointing out on this!

File: pymc/distributions/distribution.py [hunk 6]
```diff
@@ -83,6 +85,61 @@
...
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            if isinstance(node.op, Scan):
+                # inner_graph = node.op.fgraph.clone()
+                # self.rewrite(inner_graph)
+                # node.op.fgraph = inner_graph
+                new_node = rewrite_moment_scan_node(node)
+                for out1, out2 in zip(node.outputs, new_node.outputs):
+                    fgraph.replace(out1, out2)
+            elif isinstance(node.op, (RandomVariable, SymbolicRandomVariable)):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: I think this branch should come first

File: pymc/distributions/distribution.py [hunk 7]
```diff
@@ -687,6 +747,7 @@ def custom_dist_logcdf(op, value, size, *params, **kwargs):
 
         @_moment.register(rv_type)
         def custom_dist_get_moment(op, rv, size, *params):
+            params = [i for i in params if not isinstance(i, RandomGeneratorSharedVariable)]
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Why is this needed?

File: pymc/distributions/distribution.py (hunk 5 above)
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Should we specify the FunctionGraph inputs? I guess something like:

```python
rv = dist(*dis_params, size=size) 
# Filter shared variables / constants that I think cannot be inputs to the FunctionGraph
filtered_inputs = [inp for inp in rv.owner.inputs if inp...]
fgraph = FunctionGraph(inputs=filtered_inputs, outputs=rv, clone=True)
```

File: pymc/distributions/distribution.py (hunk 7 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: `params` consists `RandomGeneratorSharedVariable`s, which are not match with `dist` function signature. So I filter out it here, but may be there is a better way to do it

File: pymc/distributions/distribution.py (hunk 5 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: I experimented with this approach a little bit and it is not working. We need a whole graph, but `rv.owner.inputs` get us only last node and I got `MissingInputError`. For example for  `pt.exp(pm.Normal.dist(mu, sigma, size=size))` dist function `rv.owner.inputs` is just `[normal_rv{0, (0, 0), floatX, False}.out]`

File: pymc/distributions/distribution.py (hunk 7 above)
Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Ah, I think below we do something like `params[:len(dist_params)]`? Does that work? Would be nice to add a comment to say we are excluding the shared RNGs

File: pymc/distributions/distribution.py (hunk 5 above)
Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Hmm, that doesn't sound right. `dist` in a CustomDist should return an `OpFromGraph`, whose inputs would be `size` `mu`, `sigma`, and the shared RNGs.

```python
import pymc as pm

def dist_fn(mu, sigma, size):
    return pm.math.exp(pm.Normal.dist(mu, sigma, shape=size))

rv = pm.CustomDist.dist(0, 1, dist=dist_fn)
rv.owner.inputs
# [TensorConstant(TensorType(int64, shape=(0,)), data=array([], ... ype=int64)),
#  TensorConstant(TensorType(int8, shape=()), data=array(0, dtype=int8)),
#  TensorConstant(TensorType(int8, shape=()), data=array(1, dtype=int8)),
#  RandomGeneratorSharedVariable(<Generator(PCG64) at 0x7F61326D0820>)]
```

The point is that we don't want to rewrite the graph above the CustomDist.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Actually we can just apply the moment in the `rv.owner.op.fgraph.outputs` (after cloning it) and return that? More specifically on the output that corresponds to `rv`?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94  now we take output of `dist_fn` (not `CustomDist.dist`), build fgraph  and apply moment to it nodes. Do you think we need to take `CustomDist.dist` instead of `dist_fn`?

File: pymc/distributions/distribution.py (hunk 7 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: I think `params[:len(dist_params)]` should work here, but do we have access to `dist_params` from this function? Looks like no.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: In that case let's keep like you did, but perhaps use a helper function with a readable name? Could that helper be used below?

File: pymc/distributions/distribution.py (hunk 5 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py:678]: Will add comment

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: > @ricardoV94  now we take output of `dist_fn` (not `CustomDist.dist`), build fgraph  and apply moment to it nodes. Do you think we need to take `CustomDist.dist` instead of `dist_fn`? 

Yes that's better, because it already takes care of encapsulating the graph that corresponds to the CustomDist

Comment (aerubanov) [reply pymc/distributions/distribution.py]: I think we can not use `CustomDist.dist` because we need to build `moment` inside `dist` method

File: pymc/distributions/distribution.py (hunk 7 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: I added helper function to filter out shared RNGs

File: pymc/distributions/distribution.py (hunk 5 above)
Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Can we extract the code that creates the fgraph from user provided fn and inputs into a helper outside of `CustomDist.dist`? That was the only reason I was suggesting we call it again.

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 I found that `FunctionGraph` can extract and filter graph inputs on its own:
 ```
         if inputs is None:
            inputs = [
                i for i in graph_inputs(outputs) if not isinstance(i, AtomicVariable)
            ]
```
 https://github.com/pymc-devs/pytensor/blob/7bb18f3a3590d47132245b7868b3a4a6587a4667/pytensor/graph/fg.py#L105
 so I think we need not to specify inputs manually. But I moved graph construction in helper function anyway.

File: pymc/distributions/distribution.py [hunk 8]
```diff
@@ -83,6 +85,60 @@
 PLATFORM = sys.platform
 
 
+def filter_shared_RNGs(params):
+    return [p for p in params if not isinstance(p, RandomGeneratorSharedVariable)]
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: This should be more general and still apply to your use case.
```suggestion
def filter_RNGs(params):
    return [p for p in params if not isinstance(p.type, RandomType)]
```

File: pymc/distributions/distribution.py [hunk 9]
```diff
@@ -83,6 +85,60 @@
...
+
+
+def rewrite_moment_scan_node(node):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: This can be a method of `MomentRewriter` just to keep things more tidy

File: pymc/distributions/distribution.py [hunk 10]
```diff
@@ -622,13 +685,15 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
+        def dist_moment(rv, size, *dist_params):
+            fgraph = get_rv_fgraph(dist, dist_params, size)
+            replace_moments = MomentRewrite()
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: I am still worried that this will rewrite `dist_params` as well, and not just the CustomDist graph between `dist` and `dist_params`. 

Would the following work? Clone the inner fgraph from `rv.owner.op.fgraph`. This graph doesn't have `dist_params` directly but dummy placeholders called `NominalVariable`s. Apply the rewrite on this inner graph, and once you are done, replace the `NominalVariable`s by the respective `dist_params`?

**Review by ricardoV94 — COMMENTED** [review]
`CustomDist` is getting more complex, we might want to move it into its own file later down the road.

File: pymc/distributions/distribution.py (hunk 10 above)
Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Actually what I am suggesting is pretty similar to what you did with `Scan`, so you could perhaps reuse some of the same logic?

File: pymc/distributions/distribution.py [hunk 11]
```diff
@@ -83,6 +85,60 @@
...
+        if nd not in to_replace_set and isinstance(nd.op, (RandomVariable, SymbolicRandomVariable)):
+            replace_with_moment.append(nd.out)
+            to_replace_set.add(nd)
+    givens = {}
+    if len(replace_with_moment) > 0:
+        for item in replace_with_moment:
+            givens[item] = moment(item)
+    op_outs = clone_replace(node_outputs, replace=givens)
+
+    nwScan = Scan(
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Is there a way to check if there was actually any replacement done? If not, there's no point in creating a new Scan

Comment (aerubanov) [reply pymc/distributions/distribution.py]: Yeah, I think we can add this check

File: pymc/distributions/distribution.py (hunk 10 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94  When I trying to create `FunctionGraph` with `NominalVariable`s I get error about missing input values - I think `FunctionGraph` constructor do not support dummy placeholders as input. But I`m try another approach:
```python
def dist_moment(rv, size, *dist_params, dist):
    rv = dist(*dist_params, size=size)
    inputs, outputs = dist_params, [rv.owner.out]
    fgraph_topo = io_toposort(inputs, outputs)
    replace_with_moment = []
    to_replace_set = set()

    for nd in fgraph_topo:
          if nd not in to_replace_set and isinstance(
               nd.op, (RandomVariable, SymbolicRandomVariable)
          ):
                replace_with_moment.append(nd.out)
                to_replace_set.add(nd)
     givens = {}
     for item in replace_with_moment:
            givens[item] = moment(item)
     [out] = clone_replace(outputs, replace=givens)
     return out
```
Looks like it work but do not support `Scan` for now (but I can add it). Do you think such approach will be better?

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: You don't have to recreate the `FunctionGraph`, you can just do `fgraph.clone()`?

**Comment by ricardoV94** [comment]
BTW @aerubanov although this is a tricky one it seems to be getting really good! Great work so far

File: pymc/distributions/distribution.py (hunk 10 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 I try this approach, but I need to create `CustomDist` op first because just output of `dist` function do not have `fgraph` attribute.
```python
def dist_moment(rv, size, *dist_params):
            size = normalize_size_param(size)
            dummy_size_param = size.type()
            dummy_dist_params = [dist_param.type() for dist_param in dist_params]
            dummy_rv = dist(*dummy_dist_params, dummy_size_param)
            dummy_params = [dummy_size_param] + dummy_dist_params
            rv_type = type(
                class_name,
                (CustomSymbolicDistRV,),
                # If logp is not provided, we try to infer it from the dist graph
                dict(
                    inline_logprob=logp is None,
                ),
            )
            rv_op = rv_type(
                inputs=dummy_params,
                outputs=[dummy_rv],
                ndim_supp=ndim_supp,
            )
            fgraph = rv_op.fgraph.clone()
            replace_moments = MomentRewrite()
            replace_moments.rewrite(fgraph)
            for i, par in enumerate([size] + list(dist_params)):
                fgraph.replace(fgraph.inputs[i], par)
            [moment] = fgraph.outputs
            return moment
```
Do you think this way better? It is  also do not work with `Scan` yet. I need to figure out why and I will move graph creation logic to helper function if we will decide to keep this way.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: You have fgraph in the `rv.owner.op` from the `rv` that's passed as input.

About Scan what exactly is the problem? The approach you had written already seemed fine.  Was it failing?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: For scan I got Assertion Error:
```
        # allocate storage for intermediate computation
        for node in order:
            for r in node.inputs:
                if r not in storage_map:
>                   assert isinstance(r, Constant)
E                   AssertionError

../../miniconda3/envs/pymc-dev/lib/python3.11/site-packages/pytensor/link/utils.py:135: AssertionError
```
It is hapens when I run `TestCustomSymbolicDist::test_custom_dist_default_moment_inner_graph` test case

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 If I create rv like this it`s do not have fgraph atribute:
```
rv = dist(*dist_params, size=size)
>>> rv
Exp.0
>>> rv.owner
Exp(normal_rv{0, (0, 0), floatX, False}.out)
>>> rv.owner.op
Elemwise(scalar_op=exp,inplace_pattern=<frozendict {}>)
>>> rv.owner.op.fgraph
*** AttributeError: 'Elemwise' object has no attribute 'fgraph'
```
I use `lambda mu, sigma, size: pt.exp(pm.Normal.dist(mu, sigma, size=size))` for `dist` function  here.

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: No, use the `rv` that is passed to the moment function directly (first argument), don't recreate it

Comment (aerubanov) [reply pymc/distributions/distribution.py]: Ohm yeah, my bad. Will try it

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 No, I can not use `rv` that is passes as parameter:
```
>>> rv
*** Not yet returned!
```
So I tried to re-create it

File: pymc/distributions/distribution.py [hunk 12]
```diff
@@ -622,13 +682,38 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
-        if moment is None:
-            moment = functools.partial(
-                default_moment,
-                rv_name=class_name,
-                has_fallback=True,
+        def dist_moment(rv, size, *dist_params):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Why can't you use the fgraph of the `rv` here? I don't follow

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: Here is what I have in mind:

```python
def dist_moment(rv, *rv_inputs):
  rv_op = rv.owner.op
  fgraph = rv_op.fgraph.clone()
  replace_moments = MomentRewrite()
  replace_moments.rewrite(fgraph)
  fgraph.replace_all(tuple(zip(rv_op.inner_inputs, rv_inputs)))
  [moment] = fgraph.outputs
  return moment
```

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 because I get
```
>>> rv
*** Not yet returned!
```
Not sure why actually

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: I don't know what Not yet returned means?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 I just used pdb wrong, we do not need re-create rv here.

**Comment by aerubanov** [comment]
@ricardoV94 Could you please take a look on recent changes? Just friendly reminder)

File: pymc/distributions/distribution.py [hunk 13]
```diff
@@ -622,13 +683,28 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
+        def dist_moment(rv, size, *dist_params):
+            fgraph = rv.owner.op.fgraph.clone()
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: This method can now be used for any OpFromGraph not just CustomDist ones, so I would move it to the MomentRewriter

File: pymc/distributions/distribution.py [hunk 14]
```diff
@@ -622,13 +683,28 @@ def dist(
...
+            replace_moments = MomentRewrite()
+            replace_moments.rewrite(fgraph)
+            # we need to replace dummy variable by actual dist params
+            for i, par in enumerate([size] + list(dist_params)):
+                fgraph.replace(fgraph.inputs[i], par)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: You can use `fgraph.replace_all` to avoid the loop

File: pymc/distributions/distribution.py [hunk 15]
```diff
@@ -622,13 +683,28 @@ def dist(
...
+            # we need to replace dymmy random generators in Scan node inputs
+            for node in fgraph.toposort():
+                if isinstance(node.op, Scan):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Why do we need a special logic for Scan inside an OpFromGraph?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 After cloning fgraph we have dummy placeholders in `Scan` inner graph inputs for `RandomGenerator`, and we need to replace it by real random generator to avoid error later when we will try to calculate moment. Do we have another way to do it?

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: I see, so the problem is that the scan moment rewrite does not remove the useless updates. That's probably okay. The issue is that we filtered out the RNGs!!! If we had them the replacement would have worked fine. I was able to simplify the function to this and it passes the test that has a scan:

```python
        def dist_moment(rv, *args):
            node = rv.owner
            op = node.op
            rv_out_idx = node.outputs.index(rv)
            
            fgraph = op.fgraph.clone()
            replace_moments = MomentRewrite()
            replace_moments.rewrite(fgraph)
            # Replace dummy inner inputs by outer inputs
            fgraph.replace_all(tuple(zip(op.inner_inputs, node.inputs)), import_missing=True)
            moment = fgraph.outputs[rv_out_idx]
            return moment
```

Note that I basically got a handle of the filtered RNGs by doing node.inputs. We don't need to do this if we don't filter them out.

To tidy up the code, we should dispatch this moment function on `CustomSymbolicDistRV` outside of the `_CustomSymbolicDist` class. Then we only dispatch a user provided moment when that's given like we do now with `if logp is not None:`

You can remove the moment related stuff that now exists in `_CustomSymbolicDist.dist`

Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94 Great, thank you! Will do

File: tests/distributions/test_distribution.py [hunk 16]
```diff
@@ -430,6 +430,85 @@ def custom_dist(mu, sigma, size):
...
+                lambda mu1, sigma, size: pm.Normal.dist(mu1, sigma, size=size),
+            ),
+        ],
+    )
+    def test_custom_dist_default_moment(self, dist_params, size, expected, dist_fn):
+        with Model() as model:
+            CustomDist("x", *dist_params, dist=dist_fn, size=size)
+        assert_moment_is_expected(model, expected)
+
+    def test_custom_dist_default_moment_inner_graph(self):
```
Comment (ricardoV94) [inline tests/distributions/test_distribution.py]:
```suggestion
    def test_custom_dist_default_moment_scan(self):
```

File: tests/distributions/test_distribution.py [hunk 17]
```diff
@@ -430,6 +430,85 @@ def custom_dist(mu, sigma, size):
...
+        def scan_step(left, right):
```
Comment (ricardoV94) [inline tests/distributions/test_distribution.py:476]: Can we add a second test with recurring inputs like `step = lambda xtm1: pm.Normal.dist(xtm1)` (plus the updates of course)?

File: pymc/distributions/distribution.py [hunk 18]
```diff
@@ -83,6 +85,64 @@
...
+
+    def apply(self, fgraph):
+        for node in fgraph.toposort():
+            if isinstance(node.op, (RandomVariable, SymbolicRandomVariable)):
+                fgraph.replace(node.out, moment(node.out))
+            elif isinstance(node.op, Scan):
+                new_node = self.rewrite_moment_scan_node(node)
+                if new_node is not None:
+                    for out1, out2 in zip(node.outputs, new_node.outputs):
+                        fgraph.replace(out1, out2)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Use replace_all?

File: pymc/distributions/distribution.py [hunk 19]
```diff
@@ -687,7 +756,7 @@ def custom_dist_logcdf(op, value, size, *params, **kwargs):
 
         @_moment.register(rv_type)
         def custom_dist_get_moment(op, rv, size, *params):
-            return moment(rv, size, *params[: len(params)])
+            return moment(rv, size, *filter_RNGs(params))
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Only register this when `moment is not None`. And the default case you can register in the base class `CustomSymbolicDistRV` outside

File: pymc/distributions/distribution.py [hunk 20]
```diff
@@ -623,12 +697,7 @@ def dist(
             logcdf = default_not_implemented(class_name, "logcdf")
 
         if moment is None:
-            moment = functools.partial(
-                default_moment,
-                rv_name=class_name,
-                has_fallback=True,
-                ndim_supp=ndim_supp,
-            )
+            moment = dist_moment
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Remove this if statement (see other comment about default)

File: tests/distributions/test_distribution.py [hunk 21]
```diff
@@ -430,6 +430,104 @@ def custom_dist(mu, sigma, size):
...
+            )
+            return xs
+
+        with Model() as model:
+            CustomDist("x", dist=dist)
+        assert_moment_is_expected(model, np.array([-3, -2]))
+
+    def test_custom_dist_default_moment_scan_recurring(self):
+        def scan_step(xtm1):
+            x = pm.Normal.dist(xtm1)
```
Comment (ricardoV94) [inline tests/distributions/test_distribution.py]: Let's try something that requires the recursion to provide a correct output
```suggestion
            x = pm.Normal.dist(xtm1 + 1)
```

**Review by ricardoV94 — COMMENTED** [review]
This looks really great. Just some minor cleaning up suggestions left

File: pymc/distributions/distribution.py [hunk 22]
```diff
@@ -83,6 +85,64 @@
 PLATFORM = sys.platform
 
 
+def filter_RNGs(params):
+    return [p for p in params if not isinstance(p.type, (RandomType, RandomGeneratorType))]
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Since this ended up not being needed let's just inline it in the custom moment function

File: pymc/distributions/distribution.py [hunk 23]
```diff
@@ -601,6 +661,20 @@ def update(self, node: Node):
...
+def dist_moment(rv, *args):
+    node = rv.owner
+    op = node.op
+    rv_out_idx = node.outputs.index(rv)
+
+    fgraph = op.fgraph.clone()
+    replace_moments = MomentRewrite()
+    replace_moments.rewrite(fgraph)
+    # Replace dummy inner inputs by outer inputs
+    fgraph.replace_all(tuple(zip(op.inner_inputs, node.inputs)), import_missing=True)
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Can use `args` once you dispatch this to the default case, as we won't be filtering the RVs anymore.
I suggest calling it `*inputs`

```suggestion
    fgraph.replace_all(tuple(zip(op.inner_inputs, args)), import_missing=True)
```

File: pymc/distributions/distribution.py (hunk 20 above)
Comment (aerubanov) [reply pymc/distributions/distribution.py]: @ricardoV94  Hm, I think need this condition to avoid overriding moment provided by user, how can we avoid it without condition?

Comment (aerubanov) [reply pymc/distributions/distribution.py]: More specifically, If I try to remove  if statement `TestCustomSymbolicDist::test_custom_methods` fails

Comment (ricardoV94) [reply pymc/distributions/distribution.py]: There are two levels. We dispatch the general moment to the parent base class. Whenever a new subclass is created here and the user provided a moment we register it on the subclass (so it has preference). If the user didn't provide anything, we don't register and the parent class one will be used.

File: pymc/distributions/distribution.py [hunk 24]
```diff
@@ -696,8 +695,8 @@ def dist(
         if logcdf is None:
             logcdf = default_not_implemented(class_name, "logcdf")
 
-        if moment is None:
-            moment = dist_moment
+        # if moment is None:
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Also remove `moment = dist_moment`

File: pymc/distributions/distribution.py [hunk 25]
```diff
@@ -601,6 +656,20 @@ def update(self, node: Node):
         return updates
 
 
+def dist_moment(rv, *args):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: In case it was confusing, this is how you would register it in the base class. Then you don't need to do it everytime a `CustomDist` is called, only when the user wants to provide a non-default moment

```suggestion
@_moment.register(CustomSymbolicDistRV)
def dist_moment(op, rv, *args):
```

File: pymc/distributions/distribution.py [hunk 26]
```diff
@@ -601,6 +656,22 @@ def update(self, node: Node):
         return updates
 
 
+@_moment.register(CustomSymbolicDistRV)
+def dist_moment(rv, *args):
+    x = args[0]
```
Comment (aerubanov) [inline pymc/distributions/distribution.py]: @ricardoV94 I not very like this `args[0]` here and `args[1]` bellow, I mean magic numbers for index. Do you think we can avoid it somehow?

File: pymc/distributions/distribution.py [hunk 27]
```diff
@@ -601,6 +656,22 @@ def update(self, node: Node):
         return updates
 
 
+@_moment.register(CustomSymbolicDistRV)
+def dist_moment(rv, *args):
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: First argument in now op when you register. It also means you don't need to recover it from the node inside

```suggestion
def dist_moment(op, rv, *args):
```

Comment (aerubanov) [reply pymc/distributions/distribution.py]: Yeah, thank you!

File: pymc/distributions/distribution.py [hunk 28]
```diff
@@ -601,6 +656,21 @@ def update(self, node: Node):
         return updates
 
 
+@_moment.register(CustomSymbolicDistRV)
+def dist_moment(op, rv, *args):
+    node = rv.owner
+    op = node.op
```
Comment (ricardoV94) [inline pymc/distributions/distribution.py]: Last nitpick, we already have op from the signature
```suggestion
```

**Review by ricardoV94 — APPROVED** [review]

**Comment by ricardoV94** [comment]
Great work @aerubanov!
